File: hypatia_core/hypatia_core/semantic_validation.py
# hypatia_core/semantic_validation.py
#
# Semantic Validation tools for Hypatia compiler optimization pipeline.
# - S-expression structural validation (variable preservation, node ratio)
# - Model output equivalence testing (random inputs, tolerance comparison)
# - Optimization validation reports
# - Integration with ChecksumMode (Strict/Soft/Off)

from __future__ import annotations

from typing import Optional, Dict, List, Tuple, Union

import torch
import torch.nn as nn

from _hypatia_core import (
    validate_sexpr,
    validate_optimization_py,
    validate_model_outputs,
    optimize_ast,
)


# ============================================================================
# S-expression Validation
# ============================================================================


def validate_expr(
    input_expr: str,
    output_expr: Optional[str] = None,
) -> Dict:
    """Validate that an optimization preserves semantic correctness.

    If output_expr is not provided, the optimizer is run on input_expr
    to generate the optimized version automatically.

    Args:
        input_expr: Original S-expression string
        output_expr: Optimized S-expression (None = auto-optimize)

    Returns:
        dict with validation results:
            - is_valid: bool
            - message: str
            - input_node_count, output_node_count: int
            - node_reduction: int
            - fusions_found: str representation of fusions
            - variables_preserved: bool
    """
    if output_expr is None:
        output_expr = optimize_ast(input_expr)

    return dict(validate_optimization_py(input_expr, output_expr))


def validate_structure(
    input_expr: str,
    output_expr: str,
) -> Dict:
    """Validate structural properties of an optimization.

    Checks:
    - All variables from input are preserved in output
    - Node count ratio is reasonable (< 2x)

    Args:
        input_expr: Original S-expression
        output_expr: Optimized S-expression

    Returns:
        dict with: is_valid, message, cosine_similarity, shapes_match, etc.
    """
    return dict(validate_sexpr(input_expr, output_expr))


# ============================================================================
# Model Output Validation
# ============================================================================


def validate_models(
    original: nn.Module,
    optimized: nn.Module,
    input_shape: Tuple[int, ...],
    tolerance: float = 1e-5,
    num_samples: int = 3,
) -> Dict:
    """Validate that two PyTorch models produce equivalent outputs.

    Runs random inputs through both models and compares outputs
    using max absolute difference and cosine similarity.

    Args:
        original: Original PyTorch model
        optimized: Optimized PyTorch model
        input_shape: Shape for random test inputs, e.g. (1, 64)
        tolerance: Maximum allowed absolute difference (default: 1e-5)
        num_samples: Number of random inputs to test (default: 3)

    Returns:
        dict with:
            - is_valid: bool (max_diff <= tolerance)
            - max_diff: float
            - mean_diff: float
            - cosine_similarity: float (1.0 = identical)
            - shapes_match: bool
            - message: str
    """
    return dict(validate_model_outputs(
        original, optimized,
        list(input_shape),
        tolerance,
        num_samples,
    ))


class SemanticValidator:
    """High-level semantic validation for the Hypatia optimization pipeline.

    Provides both S-expression level and model-level validation,
    with configurable tolerance and reporting.

    Example::

        validator = SemanticValidator(tolerance=1e-4)

        # Validate an S-expression optimization
        result = validator.validate_expr("(relu (linear w b x))")
        print(result["message"])

        # Validate two PyTorch models
        original = nn.Sequential(nn.Linear(64, 32), nn.ReLU())
        optimized = nn.Sequential(nn.Linear(64, 32), nn.ReLU())
        result = validator.validate_models(original, optimized, input_shape=(1, 64))
        assert result["is_valid"]
    """

    def __init__(
        self,
        tolerance: float = 1e-5,
        num_samples: int = 5,
        strict: bool = False,
    ):
        """Initialize the semantic validator.

        Args:
            tolerance: Maximum absolute difference for output comparison
            num_samples: Number of random test inputs
            strict: If True, also validate structural properties
        """
        self.tolerance = tolerance
        self.num_samples = num_samples
        self.strict = strict

    def validate_expr(
        self,
        input_expr: str,
        output_expr: Optional[str] = None,
    ) -> Dict:
        """Validate an S-expression optimization.

        Args:
            input_expr: Original expression
            output_expr: Optimized expression (None = auto-optimize)

        Returns:
            Validation result dict
        """
        return validate_expr(input_expr, output_expr)

    def validate_models(
        self,
        original: nn.Module,
        optimized: nn.Module,
        input_shape: Tuple[int, ...],
    ) -> Dict:
        """Validate model output equivalence.

        Args:
            original: Original model
            optimized: Optimized model
            input_shape: Input tensor shape

        Returns:
            Validation result dict
        """
        return validate_models(
            original, optimized,
            input_shape,
            tolerance=self.tolerance,
            num_samples=self.num_samples,
        )

    def full_validation(
        self,
        input_expr: str,
        original: Optional[nn.Module] = None,
        optimized: Optional[nn.Module] = None,
        input_shape: Optional[Tuple[int, ...]] = None,
    ) -> Dict:
        """Run comprehensive validation: structural + model output.

        Args:
            input_expr: Original S-expression
            original: Original PyTorch model (optional)
            optimized: Optimized PyTorch model (optional)
            input_shape: Input tensor shape (required if models provided)

        Returns:
            Combined validation report
        """
        # Step 1: S-expression structural validation
        expr_result = self.validate_expr(input_expr)

        result = {
            "expr_valid": expr_result.get("is_valid", "true") == "true",
            "expr_message": expr_result.get("message", ""),
            "node_reduction": expr_result.get("node_reduction", "0"),
            "fusions_found": expr_result.get("fusions_found", "[]"),
            "variables_preserved": expr_result.get("variables_preserved", "true") == "true",
        }

        # Step 2: Model output validation (if models provided)
        if original is not None and optimized is not None and input_shape is not None:
            model_result = self.validate_models(original, optimized, input_shape)
            result["model_valid"] = model_result["is_valid"]
            result["max_diff"] = model_result["max_diff"]
            result["mean_diff"] = model_result["mean_diff"]
            result["cosine_similarity"] = model_result["cosine_similarity"]
            result["model_message"] = model_result["message"]
            result["is_valid"] = result["expr_valid"] and model_result["is_valid"]
        else:
            result["is_valid"] = result["expr_valid"]

        return result

    def report(
        self,
        input_expr: str,
        original: Optional[nn.Module] = None,
        optimized: Optional[nn.Module] = None,
        input_shape: Optional[Tuple[int, ...]] = None,
    ) -> str:
        """Generate a human-readable validation report.

        Args:
            input_expr: Original S-expression
            original: Original PyTorch model (optional)
            optimized: Optimized PyTorch model (optional)
            input_shape: Input tensor shape (optional)

        Returns:
            Formatted report string
        """
        result = self.full_validation(input_expr, original, optimized, input_shape)

        lines = []
        lines.append("=" * 60)
        lines.append("  HYPATIA SEMANTIC VALIDATION REPORT")
        lines.append("=" * 60)
        lines.append("")

        # Overall status
        status = "PASSED" if result["is_valid"] else "FAILED"
        lines.append(f"  Status: {status}")
        lines.append(f"  Tolerance: {self.tolerance}")
        lines.append(f"  Test samples: {self.num_samples}")
        lines.append("")

        # Expression validation
        lines.append("  Expression Validation:")
        lines.append(f"    Valid: {result['expr_valid']}")
        lines.append(f"    Message: {result['expr_message']}")
        lines.append(f"    Variables preserved: {result['variables_preserved']}")
        lines.append(f"    Node reduction: {result['node_reduction']}")
        lines.append(f"    Fusions: {result['fusions_found']}")
        lines.append("")

        # Model validation (if available)
        if "model_valid" in result:
            lines.append("  Model Output Validation:")
            lines.append(f"    Valid: {result['model_valid']}")
            lines.append(f"    Max diff: {result['max_diff']:.2e}")
            lines.append(f"    Mean diff: {result['mean_diff']:.2e}")
            lines.append(f"    Cosine similarity: {result['cosine_similarity']:.6f}")
            lines.append(f"    Message: {result['model_message']}")
            lines.append("")

        lines.append("=" * 60)

        report_str = "\n".join(lines)
        print(report_str)
        return report_str


__all__ = [
    "validate_expr",
    "validate_structure",
    "validate_models",
    "SemanticValidator",
]

File: hypatia_core/src/semantic_validation.rs
//! Semantic Validation: Verify optimization correctness via output comparison
//!
//! Provides runtime validation that optimized models produce equivalent outputs:
//! - Random input generation with shape inference from model parameters
//! - Forward pass comparison with configurable tolerance
//! - Shape matching validation
//! - Statistical divergence analysis (max diff, mean diff, cosine similarity)
//!
//! Integrates with ChecksumMode:
//! - Strict: Full semantic validation with random inputs
//! - Soft: Shape validation + lightweight output comparison
//! - Off: No validation

use std::collections::HashMap;

// ============================================================================
// Validation Result
// ============================================================================

/// Result of semantic validation between original and optimized models.
#[derive(Debug, Clone)]
pub struct ValidationResult {
    /// Whether the models are semantically equivalent within tolerance.
    pub is_valid: bool,
    /// Maximum absolute difference between outputs.
    pub max_diff: f64,
    /// Mean absolute difference between outputs.
    pub mean_diff: f64,
    /// Cosine similarity between flattened outputs (1.0 = identical).
    pub cosine_similarity: f64,
    /// Whether output shapes match.
    pub shapes_match: bool,
    /// Original model output shape.
    pub original_shape: Vec<i64>,
    /// Optimized model output shape.
    pub optimized_shape: Vec<i64>,
    /// Number of test inputs used.
    pub num_test_inputs: usize,
    /// Tolerance used for comparison.
    pub tolerance: f64,
    /// Human-readable validation message.
    pub message: String,
}

// ============================================================================
// Pure Rust Utilities (no Python dependency)
// ============================================================================

/// Compute cosine similarity between two f64 slices.
pub fn cosine_similarity(a: &[f64], b: &[f64]) -> f64 {
    if a.len() != b.len() || a.is_empty() {
        return 0.0;
    }

    let mut dot = 0.0;
    let mut norm_a = 0.0;
    let mut norm_b = 0.0;

    for i in 0..a.len() {
        dot += a[i] * b[i];
        norm_a += a[i] * a[i];
        norm_b += b[i] * b[i];
    }

    let denom = norm_a.sqrt() * norm_b.sqrt();
    if denom < 1e-12 {
        // Both near-zero → consider identical
        if norm_a < 1e-12 && norm_b < 1e-12 {
            return 1.0;
        }
        return 0.0;
    }

    dot / denom
}

/// Compute max and mean absolute difference between two f64 slices.
pub fn compute_diffs(a: &[f64], b: &[f64]) -> (f64, f64) {
    if a.len() != b.len() || a.is_empty() {
        return (f64::INFINITY, f64::INFINITY);
    }

    let mut max_diff = 0.0f64;
    let mut sum_diff = 0.0f64;

    for i in 0..a.len() {
        let diff = (a[i] - b[i]).abs();
        max_diff = max_diff.max(diff);
        sum_diff += diff;
    }

    let mean_diff = sum_diff / a.len() as f64;
    (max_diff, mean_diff)
}

/// Build a ValidationResult from computed metrics.
pub fn build_validation_result(
    original_values: &[f64],
    optimized_values: &[f64],
    original_shape: Vec<i64>,
    optimized_shape: Vec<i64>,
    num_test_inputs: usize,
    tolerance: f64,
) -> ValidationResult {
    let shapes_match = original_shape == optimized_shape;

    if !shapes_match {
        return ValidationResult {
            is_valid: false,
            max_diff: f64::INFINITY,
            mean_diff: f64::INFINITY,
            cosine_similarity: 0.0,
            shapes_match: false,
            original_shape,
            optimized_shape,
            num_test_inputs,
            tolerance,
            message: "Output shapes do not match".to_string(),
        };
    }

    let (max_diff, mean_diff) = compute_diffs(original_values, optimized_values);
    let cos_sim = cosine_similarity(original_values, optimized_values);

    let is_valid = max_diff <= tolerance;

    let message = if is_valid {
        format!(
            "Semantic validation PASSED: max_diff={:.2e} <= tolerance={:.2e}, cosine_sim={:.6}",
            max_diff, tolerance, cos_sim
        )
    } else {
        format!(
            "Semantic validation FAILED: max_diff={:.2e} > tolerance={:.2e}, cosine_sim={:.6}",
            max_diff, tolerance, cos_sim
        )
    };

    ValidationResult {
        is_valid,
        max_diff,
        mean_diff,
        cosine_similarity: cos_sim,
        shapes_match,
        original_shape,
        optimized_shape,
        num_test_inputs,
        tolerance,
        message,
    }
}

// ============================================================================
// S-expression Level Validation
// ============================================================================

/// Validate that an optimized S-expression is semantically equivalent
/// to the original by checking structural properties.
///
/// This is a lightweight check that doesn't require running models:
/// - Node count should not increase dramatically
/// - All variable names from input should appear in output
/// - No unknown operators should be introduced
pub fn validate_sexpr_structure(
    input_expr: &str,
    output_expr: &str,
) -> Result<ValidationResult, String> {
    use crate::egraph_optimizer::HypatiaLang;
    use egg::RecExpr;

    let input_parsed: RecExpr<HypatiaLang> = input_expr
        .parse()
        .map_err(|e| format!("Parse input error: {}", e))?;
    let output_parsed: RecExpr<HypatiaLang> = output_expr
        .parse()
        .map_err(|e| format!("Parse output error: {}", e))?;

    let input_count = input_parsed.as_ref().len();
    let output_count = output_parsed.as_ref().len();

    // Extract variable names from both expressions
    let input_vars = extract_variables(&input_parsed);
    let output_vars = extract_variables(&output_parsed);

    // Check that all input variables appear in output
    let mut missing_vars = Vec::new();
    for var in &input_vars {
        if !output_vars.contains(var) {
            missing_vars.push(var.clone());
        }
    }

    // Node reduction ratio (optimization shouldn't increase nodes dramatically)
    let node_ratio = output_count as f64 / input_count.max(1) as f64;

    let is_valid = missing_vars.is_empty() && node_ratio <= 2.0;

    let message = if is_valid {
        format!(
            "Structural validation PASSED: {} → {} nodes (ratio: {:.2}), all {} variables preserved",
            input_count, output_count, node_ratio, input_vars.len()
        )
    } else if !missing_vars.is_empty() {
        format!(
            "Structural validation FAILED: variables lost: {:?}",
            missing_vars
        )
    } else {
        format!(
            "Structural validation FAILED: node count increased too much ({} → {}, ratio: {:.2})",
            input_count, output_count, node_ratio
        )
    };

    Ok(ValidationResult {
        is_valid,
        max_diff: 0.0,
        mean_diff: 0.0,
        cosine_similarity: if is_valid { 1.0 } else { 0.0 },
        shapes_match: true,
        original_shape: vec![input_count as i64],
        optimized_shape: vec![output_count as i64],
        num_test_inputs: 0,
        tolerance: 0.0,
        message,
    })
}

/// Extract all variable names from a RecExpr.
fn extract_variables(expr: &egg::RecExpr<crate::egraph_optimizer::HypatiaLang>) -> Vec<String> {
    use crate::egraph_optimizer::HypatiaLang;

    let mut vars = Vec::new();
    for node in expr.as_ref() {
        if let HypatiaLang::Var(s) = node {
            let name = s.to_string();
            if !vars.contains(&name) {
                vars.push(name);
            }
        }
    }
    vars
}

/// Validate optimization and return a structured report.
/// Combines structural validation with the optimization report.
pub fn validate_optimization(
    input_expr: &str,
    output_expr: &str,
) -> Result<HashMap<String, String>, String> {
    let struct_result = validate_sexpr_structure(input_expr, output_expr)?;
    let opt_report = crate::visualization::build_optimization_report(input_expr, output_expr)?;

    let mut report = HashMap::new();
    report.insert("is_valid".to_string(), struct_result.is_valid.to_string());
    report.insert("message".to_string(), struct_result.message.clone());
    report.insert("input_expr".to_string(), input_expr.to_string());
    report.insert("output_expr".to_string(), output_expr.to_string());
    report.insert("input_node_count".to_string(), opt_report.input_node_count.to_string());
    report.insert("output_node_count".to_string(), opt_report.output_node_count.to_string());
    report.insert("node_reduction".to_string(), opt_report.node_reduction.to_string());
    report.insert(
        "fusions_found".to_string(),
        format!("{:?}", opt_report.fusions_found),
    );
    report.insert(
        "variables_preserved".to_string(),
        {
            let input_parsed: egg::RecExpr<crate::egraph_optimizer::HypatiaLang> =
                input_expr.parse().unwrap();
            let output_parsed: egg::RecExpr<crate::egraph_optimizer::HypatiaLang> =
                output_expr.parse().unwrap();
            let input_vars = extract_variables(&input_parsed);
            let output_vars = extract_variables(&output_parsed);
            let all_preserved = input_vars.iter().all(|v| output_vars.contains(v));
            all_preserved.to_string()
        },
    );

    Ok(report)
}

// ============================================================================
// Unit Tests
// ============================================================================

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_cosine_similarity_identical() {
        let a = vec![1.0, 2.0, 3.0];
        let b = vec![1.0, 2.0, 3.0];
        let sim = cosine_similarity(&a, &b);
        assert!((sim - 1.0).abs() < 1e-10);
    }

    #[test]
    fn test_cosine_similarity_orthogonal() {
        let a = vec![1.0, 0.0];
        let b = vec![0.0, 1.0];
        let sim = cosine_similarity(&a, &b);
        assert!(sim.abs() < 1e-10);
    }

    #[test]
    fn test_cosine_similarity_opposite() {
        let a = vec![1.0, 2.0, 3.0];
        let b = vec![-1.0, -2.0, -3.0];
        let sim = cosine_similarity(&a, &b);
        assert!((sim + 1.0).abs() < 1e-10);
    }

    #[test]
    fn test_cosine_similarity_zero() {
        let a = vec![0.0, 0.0, 0.0];
        let b = vec![0.0, 0.0, 0.0];
        let sim = cosine_similarity(&a, &b);
        assert!((sim - 1.0).abs() < 1e-10); // Both zero → identical
    }

    #[test]
    fn test_compute_diffs_identical() {
        let a = vec![1.0, 2.0, 3.0];
        let b = vec![1.0, 2.0, 3.0];
        let (max_diff, mean_diff) = compute_diffs(&a, &b);
        assert!(max_diff < 1e-10);
        assert!(mean_diff < 1e-10);
    }

    #[test]
    fn test_compute_diffs_known() {
        let a = vec![1.0, 2.0, 3.0];
        let b = vec![1.1, 2.0, 3.3];
        let (max_diff, mean_diff) = compute_diffs(&a, &b);
        assert!((max_diff - 0.3).abs() < 1e-10);
        assert!((mean_diff - (0.1 + 0.0 + 0.3) / 3.0).abs() < 1e-10);
    }

    #[test]
    fn test_build_validation_result_pass() {
        let a = vec![1.0, 2.0, 3.0];
        let b = vec![1.0 + 1e-7, 2.0, 3.0 - 1e-8];
        let result = build_validation_result(
            &a, &b,
            vec![1, 3], vec![1, 3],
            1, 1e-5,
        );
        assert!(result.is_valid);
        assert!(result.shapes_match);
        assert!(result.max_diff < 1e-5);
    }

    #[test]
    fn test_build_validation_result_fail() {
        let a = vec![1.0, 2.0, 3.0];
        let b = vec![1.5, 2.0, 3.0];
        let result = build_validation_result(
            &a, &b,
            vec![1, 3], vec![1, 3],
            1, 1e-5,
        );
        assert!(!result.is_valid);
        assert!((result.max_diff - 0.5).abs() < 1e-10);
    }

    #[test]
    fn test_build_validation_result_shape_mismatch() {
        let a = vec![1.0, 2.0];
        let b = vec![1.0, 2.0, 3.0];
        let result = build_validation_result(
            &a, &b,
            vec![1, 2], vec![1, 3],
            1, 1e-5,
        );
        assert!(!result.is_valid);
        assert!(!result.shapes_match);
    }

    #[test]
    fn test_validate_sexpr_structure_basic() {
        let result = validate_sexpr_structure(
            "(relu (linear w b x))",
            "(fused_linear_relu w b x)",
        ).unwrap();
        assert!(result.is_valid);
    }

    #[test]
    fn test_validate_sexpr_structure_same() {
        let result = validate_sexpr_structure(
            "(relu x)",
            "(relu x)",
        ).unwrap();
        assert!(result.is_valid);
    }

    #[test]
    fn test_validate_sexpr_structure_variable_preserved() {
        let result = validate_sexpr_structure(
            "(relu (linear w b x))",
            "(fused_linear_relu w b x)",
        ).unwrap();
        assert!(result.is_valid);
        assert!(result.message.contains("PASSED"));
    }

    #[test]
    fn test_validate_optimization_report() {
        let report = validate_optimization(
            "(relu (linear w b x))",
            "(fused_linear_relu w b x)",
        ).unwrap();
        assert_eq!(report["is_valid"], "true");
        assert_eq!(report["variables_preserved"], "true");
    }

    #[test]
    fn test_validate_optimization_no_change() {
        let report = validate_optimization(
            "(relu x)",
            "(relu x)",
        ).unwrap();
        assert_eq!(report["is_valid"], "true");
        assert_eq!(report["node_reduction"], "0");
    }

    #[test]
    fn test_extract_variables() {
        let expr: egg::RecExpr<crate::egraph_optimizer::HypatiaLang> =
            "(relu (linear w b x))".parse().unwrap();
        let vars = extract_variables(&expr);
        assert!(vars.contains(&"w".to_string()));
        assert!(vars.contains(&"b".to_string()));
        assert!(vars.contains(&"x".to_string()));
        assert_eq!(vars.len(), 3);
    }
}
